File: frontend/src/pkg/mermaid/metadata.ts
import type { PlanConnector, PlanElement, TechnologyLink } from '@buf/tldiagramcom_diagram.bufbuild_es/diag/v1/workspace_service_pb'

interface TldElementMetadata {
  ref?: string
  x?: number
  y?: number
  kind?: string
  description?: string
  technology?: string
  url?: string
  logoUrl?: string
  tags?: string[]
  technologyLinks?: TechnologyLink[]
  repo?: string
  branch?: string
  filePath?: string
  language?: string
  bypassNoiseGate?: boolean
  hasView?: boolean
  viewLabel?: string
}

interface TldConnectorMetadata {
  ref?: string
  sourceRef?: string
  targetRef?: string
  label?: string
  description?: string
  relationship?: string
  direction?: string
  style?: string
  url?: string
  sourceHandle?: string
  targetHandle?: string
}

interface TldMetadata {
  elements: Map<string, TldElementMetadata>
  connectors: TldConnectorMetadata[]
}

interface MetadataImportResult {
  elements: PlanElement[]
  connectors: PlanConnector[]
}

export interface TldTechnologyLinkExport {
  type: 'catalog' | 'custom'
  slug?: string
  label: string
  is_primary_icon?: boolean
  isPrimaryIcon?: boolean
}

function splitEscaped(value: string, separator: string) {
  const parts: string[] = []
  let current = ''
  for (let index = 0; index < value.length; index += 1) {
    const char = value[index]
    if (char === '\\' && index + 1 < value.length) {
      current += char + value[index + 1]
      index += 1
      continue
    }
    if (char === separator) {
      parts.push(current)
      current = ''
      continue
    }
    current += char
  }
  parts.push(current)
  return parts
}

export function escapeTldMetadataValue(value: string) {
  return value
    .replace(/\\/g, '\\\\')
    .replace(/\n/g, '\\n')
    .replace(/\r/g, '\\r')
    .replace(/\t/g, '\\t')
    .replace(/=/g, '\\=')
    .replace(/,/g, '\\,')
    .replace(/\|/g, '\\|')
    .replace(/:/g, '\\:')
}

function unescapeTldMetadataValue(value: string) {
  let out = ''
  for (let index = 0; index < value.length; index += 1) {
    const char = value[index]
    if (char !== '\\') {
      out += char
      continue
    }
    index += 1
    if (index >= value.length) return null
    switch (value[index]) {
      case '\\':
      case '=':
      case ',':
      case '|':
      case ':':
        out += value[index]
        break
      case 'n':
        out += '\n'
        break
      case 'r':
        out += '\r'
        break
      case 't':
        out += '\t'
        break
      default:
        return null
    }
  }
  return out
}

export function encodeTldStringList(values: string[]) {
  return values.map(escapeTldMetadataValue).join(',')
}

function decodeTldStringList(value: string) {
  if (!value) return []
  const out: string[] = []
  for (const raw of splitEscaped(value, ',')) {
    const item = unescapeTldMetadataValue(raw)
    if (item === null) return null
    out.push(item)
  }
  return out
}

function technologyLinkIsPrimary(link: TldTechnologyLinkExport) {
  return Boolean(link.is_primary_icon ?? link.isPrimaryIcon)
}

export function encodeTldTechnologyLinks(values: TldTechnologyLinkExport[]) {
  return values.map((link) => [
    link.type,
    link.slug ?? '',
    link.label,
    technologyLinkIsPrimary(link) ? '1' : '0',
  ].map(escapeTldMetadataValue).join(':')).join('|')
}

function decodeTldTechnologyLinks(value: string): TechnologyLink[] | null {
  if (!value) return []
  const links: TechnologyLink[] = []
  for (const rawLink of splitEscaped(value, '|')) {
    const fields = splitEscaped(rawLink, ':')
    if (fields.length < 3 || fields.length > 4) return null
    const [rawType, rawSlug, rawLabel, rawPrimary = '0'] = fields
    const type = unescapeTldMetadataValue(rawType)
    const slug = unescapeTldMetadataValue(rawSlug)
    const label = unescapeTldMetadataValue(rawLabel)
    const primary = unescapeTldMetadataValue(rawPrimary)
    if ((type !== 'catalog' && type !== 'custom') || slug === null || label === null || primary === null) return null
    if (!label.trim()) return null
    links.push({
      type,
      slug: slug || undefined,
      label,
      isPrimaryIcon: primary === '1' || primary === 'true',
    } as TechnologyLink)
  }
  return links
}

export function compactTldNumber(value: number) {
  return Number.isInteger(value) ? String(value) : String(Math.round(value * 1000) / 1000)
}

export function finiteTldNumber(value: unknown) {
  const number = typeof value === 'number' ? value : Number(value)
  return Number.isFinite(number) ? number : null
}

function decodeTldNumber(value: string) {
  const raw = unescapeTldMetadataValue(value)
  if (raw === null) return null
  return finiteTldNumber(raw)
}

function decodeTldBoolean(value: string) {
  const raw = unescapeTldMetadataValue(value)
  if (raw === null) return null
  if (raw === '1' || raw === 'true') return true
  if (raw === '0' || raw === 'false') return false
  return null
}

function parseMetadataPairs(text: string) {
  const pairs = new Map<string, string>()
  const matches = Array.from(text.matchAll(/(^|\s+)([A-Za-z][A-Za-z0-9_]*)=/g))
  for (let index = 0; index < matches.length; index += 1) {
    const match = matches[index]
    const key = match[2]
    const keyStart = match.index + match[1].length
    const valueStart = keyStart + key.length + 1
    const nextMatch = matches[index + 1]
    const valueEnd = nextMatch?.index ?? text.length
    pairs.set(key, text.slice(valueStart, valueEnd))
  }
  return pairs
}

function parseElementMetadata(pairs: Map<string, string>): TldElementMetadata {
  const metadata: TldElementMetadata = {}
  const stringKeys: Array<[string, keyof TldElementMetadata]> = [
    ['ref', 'ref'],
    ['kind', 'kind'],
    ['desc', 'description'],
    ['tech', 'technology'],
    ['url', 'url'],
    ['logo', 'logoUrl'],
    ['repo', 'repo'],
    ['branch', 'branch'],
    ['file', 'filePath'],
    ['lang', 'language'],
    ['viewLabel', 'viewLabel'],
  ]

  for (const [key, field] of stringKeys) {
    const value = pairs.get(key)
    if (value === undefined) continue
    const decoded = unescapeTldMetadataValue(value)
    if (decoded) metadata[field] = decoded as never
  }

  const x = pairs.get('x')
  const y = pairs.get('y')
  if (x !== undefined) {
    const decoded = decodeTldNumber(x)
    if (decoded !== null) metadata.x = decoded
  }
  if (y !== undefined) {
    const decoded = decodeTldNumber(y)
    if (decoded !== null) metadata.y = decoded
  }

  const tags = pairs.get('tags')
  if (tags !== undefined) {
    const decoded = decodeTldStringList(tags)
    if (decoded && decoded.length > 0) metadata.tags = decoded
  }

  const technologyLinks = pairs.get('techLinks')
  if (technologyLinks !== undefined) {
    const decoded = decodeTldTechnologyLinks(technologyLinks)
    if (decoded && decoded.length > 0) metadata.technologyLinks = decoded
  }

  const bypass = pairs.get('bypass')
  if (bypass !== undefined) {
    const decoded = decodeTldBoolean(bypass)
    if (decoded !== null) metadata.bypassNoiseGate = decoded
  }

  const hasView = pairs.get('hasView')
  if (hasView !== undefined) {
    const decoded = decodeTldBoolean(hasView)
    if (decoded !== null) metadata.hasView = decoded
  }

  return metadata
}

function parseConnectorMetadata(pairs: Map<string, string>): TldConnectorMetadata {
  const metadata: TldConnectorMetadata = {}
  const stringKeys: Array<[string, keyof TldConnectorMetadata]> = [
    ['ref', 'ref'],
    ['source', 'sourceRef'],
    ['target', 'targetRef'],
    ['label', 'label'],
    ['desc', 'description'],
    ['rel', 'relationship'],
    ['dir', 'direction'],
    ['style', 'style'],
    ['url', 'url'],
    ['sourceHandle', 'sourceHandle'],
    ['targetHandle', 'targetHandle'],
  ]

  for (const [key, field] of stringKeys) {
    const value = pairs.get(key)
    if (value === undefined) continue
    const decoded = unescapeTldMetadataValue(value)
    if (decoded) metadata[field] = decoded
  }

  return metadata
}

function mergeConnectorMetadata(
  base: TldConnectorMetadata | undefined,
  override: TldConnectorMetadata | undefined,
): TldConnectorMetadata | undefined {
  if (!base) return override
  if (!override) return base
  return { ...base, ...override }
}

function isExportedFlowchartConnectorLine(line: string) {
  return /^[A-Za-z_][A-Za-z0-9_]*\s+(?:--\s+"(?:\\.|[^"\\])*"\s+-->|-->)\s+[A-Za-z_][A-Za-z0-9_]*$/.test(line)
}

function exportedFlowchartElementRef(line: string) {
  return line.match(/^([A-Za-z_][A-Za-z0-9_]*)\s*\["(?:\\.|[^"\\])*"\]$/)?.[1] ?? null
}

export function parseTldMetadata(source: string): TldMetadata | null {
  const lines = source.split(/\r?\n/)
  const metadata: TldMetadata = {
    elements: new Map(),
    connectors: [],
  }
  let markerSeen = false
  let lastElementRef: string | null = null
  let lastConnectorIndex = -1

  for (const line of lines) {
    const trimmed = line.trim()
    if (!trimmed.startsWith('%%')) {
      if (markerSeen) {
        lastElementRef = exportedFlowchartElementRef(trimmed)
        if (isExportedFlowchartConnectorLine(trimmed)) {
          metadata.connectors.push({})
          lastConnectorIndex = metadata.connectors.length - 1
          lastElementRef = null
        }
      }
      continue
    }
    const body = trimmed.slice(2).trim()
    if (/^tld\/v1(?:\s|$)/.test(body)) {
      markerSeen = true
      lastElementRef = null
      lastConnectorIndex = -1
      continue
    }
    if (!markerSeen) continue

    const metadataLine = body.match(/^(tld-element|tld-connector)(?:\s+(.*))?$/)
    if (!metadataLine) continue
    const [, kind, metadataText = ''] = metadataLine
    const pairs = parseMetadataPairs(metadataText)
    if (pairs.size === 0) continue
    if (kind === 'tld-element') {
      if (!lastElementRef) continue
      metadata.elements.set(lastElementRef, parseElementMetadata(pairs))
    } else if (kind === 'tld-connector') {
      const item = parseConnectorMetadata(pairs)
      if (lastConnectorIndex >= 0) {
        metadata.connectors[lastConnectorIndex] = mergeConnectorMetadata(metadata.connectors[lastConnectorIndex], item) ?? {}
      } else {
        metadata.connectors.push(item)
      }
    }
  }

  return markerSeen ? metadata : null
}

export function stripMermaidCommentLines(source: string) {
  return source
    .split(/\r?\n/)
    .filter((line) => !line.trim().startsWith('%%'))
    .join('\n')
    .trim()
}

export function applyTldMetadata(result: MetadataImportResult, metadata: TldMetadata | null) {
  if (!metadata) return

  const explicitRefs = new Map<string, string>()
  for (const element of result.elements) {
    const item = metadata.elements.get(element.ref)
    if (!item) continue
    if (item.ref && item.ref !== element.ref) {
      explicitRefs.set(element.ref, item.ref)
      element.ref = item.ref
    }
    if (item.kind) element.kind = item.kind
    if (item.description) element.description = item.description
    if (item.technology) element.technology = item.technology
    if (item.url) element.url = item.url
    if (item.logoUrl) element.logoUrl = item.logoUrl
    if (item.tags) element.tags = item.tags
    if (item.technologyLinks) element.technologyLinks = item.technologyLinks
    if (item.repo) element.repo = item.repo
    if (item.branch) element.branch = item.branch
    if (item.filePath) element.filePath = item.filePath
    if (item.language) element.language = item.language
    if (item.bypassNoiseGate !== undefined) element.bypassNoiseGate = item.bypassNoiseGate
    if (item.hasView !== undefined) element.hasView = item.hasView
    if (item.viewLabel) element.viewLabel = item.viewLabel
    if (item.x !== undefined && item.y !== undefined) {
      const placement = element.placements?.[0] ?? { parentRef: 'root' }
      placement.parentRef = placement.parentRef || 'root'
      placement.positionX = item.x
      placement.positionY = item.y
      element.placements = [placement]
    }
  }

  if (explicitRefs.size > 0) {
    for (const connector of result.connectors) {
      connector.sourceElementRef = explicitRefs.get(connector.sourceElementRef) ?? connector.sourceElementRef
      connector.targetElementRef = explicitRefs.get(connector.targetElementRef) ?? connector.targetElementRef
      connector.viewRef = explicitRefs.get(connector.viewRef) ?? connector.viewRef
    }
  }

  result.connectors.forEach((connector, index) => {
    const item = metadata.connectors[index]
    if (!item) return
    if (item.ref) connector.ref = item.ref
    if (item.sourceRef) connector.sourceElementRef = item.sourceRef
    if (item.targetRef) connector.targetElementRef = item.targetRef
    if (item.label) connector.label = item.label
    if (item.description) connector.description = item.description
    if (item.relationship) connector.relationship = item.relationship
    if (item.direction) connector.direction = item.direction
    if (item.style) connector.style = item.style
    if (item.url) connector.url = item.url
    if (item.sourceHandle) connector.sourceHandle = item.sourceHandle
    if (item.targetHandle) connector.targetHandle = item.targetHandle
  })
}

File: internal/mermaid/metadata.go
package mermaid

import (
	"regexp"
	"strconv"
	"strings"

	diagv1 "buf.build/gen/go/tldiagramcom/diagram/protocolbuffers/go/diag/v1"
)

type elementMetadata struct {
	Ref             string
	X               *float64
	Y               *float64
	Kind            string
	Description     string
	Technology      string
	URL             string
	LogoURL         string
	Tags            []string
	TechnologyLinks []*diagv1.TechnologyLink
	Repo            string
	Branch          string
	FilePath        string
	Language        string
	BypassNoiseGate *bool
	HasView         *bool
	ViewLabel       string
}

type connectorMetadata struct {
	Ref          string
	SourceRef    string
	TargetRef    string
	Label        string
	Description  string
	Relationship string
	Direction    string
	Style        string
	URL          string
	SourceHandle string
	TargetHandle string
}

type tldMetadata struct {
	Elements   map[string]elementMetadata
	Connectors []connectorMetadata
}

var (
	metadataPairPattern               = regexp.MustCompile(`(^|\s+)([A-Za-z][A-Za-z0-9_]*)=`)
	exportedFlowchartConnectorLineRe  = regexp.MustCompile(`^` + mermaidFlowchartIDPattern + `\s+(?:--\s+"(?:\\.|[^"\\])*"\s+-->|-->)\s+` + mermaidFlowchartIDPattern + `$`)
	exportedFlowchartElementRefLineRe = regexp.MustCompile(`^(` + mermaidFlowchartIDPattern + `)\s*\["(?:\\.|[^"\\])*"\]$`)
)

func ParseTldMetadata(source string) *tldMetadata {
	lines := strings.Split(strings.ReplaceAll(source, "\r\n", "\n"), "\n")
	metadata := &tldMetadata{
		Elements: map[string]elementMetadata{},
	}
	markerSeen := false
	lastElementRef := ""
	lastConnectorIndex := -1

	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if !strings.HasPrefix(trimmed, "%%") {
			if markerSeen {
				lastElementRef = exportedFlowchartElementRef(trimmed)
				if exportedFlowchartConnectorLineRe.MatchString(trimmed) {
					metadata.Connectors = append(metadata.Connectors, connectorMetadata{})
					lastConnectorIndex = len(metadata.Connectors) - 1
					lastElementRef = ""
				}
			}
			continue
		}

		body := strings.TrimSpace(strings.TrimPrefix(trimmed, "%%"))
		if regexp.MustCompile(`^tld/v1(?:\s|$)`).MatchString(body) {
			markerSeen = true
			lastElementRef = ""
			lastConnectorIndex = -1
			continue
		}
		if !markerSeen {
			continue
		}

		kind, text, ok := strings.Cut(body, " ")
		if !ok {
			kind = body
			text = ""
		}
		if kind != "tld-element" && kind != "tld-connector" {
			continue
		}
		pairs := parseMetadataPairs(text)
		if len(pairs) == 0 {
			continue
		}
		if kind == "tld-element" {
			if lastElementRef == "" {
				continue
			}
			metadata.Elements[lastElementRef] = parseElementMetadata(pairs)
			continue
		}
		item := parseConnectorMetadata(pairs)
		if lastConnectorIndex >= 0 {
			metadata.Connectors[lastConnectorIndex] = mergeConnectorMetadata(metadata.Connectors[lastConnectorIndex], item)
		} else {
			metadata.Connectors = append(metadata.Connectors, item)
		}
	}

	if !markerSeen {
		return nil
	}
	return metadata
}

func ApplyTldMetadata(parsed *ParsedDiagram, metadata *tldMetadata) {
	if metadata == nil || parsed == nil {
		return
	}

	explicitRefs := map[string]string{}
	for _, element := range parsed.Elements {
		item, ok := metadata.Elements[element.GetRef()]
		if !ok {
			continue
		}
		if item.Ref != "" && item.Ref != element.GetRef() {
			explicitRefs[element.GetRef()] = item.Ref
			element.Ref = item.Ref
		}
		if item.Kind != "" {
			element.Kind = &item.Kind
		}
		if item.Description != "" {
			element.Description = &item.Description
		}
		if item.Technology != "" {
			element.Technology = &item.Technology
		}
		if item.URL != "" {
			element.Url = &item.URL
		}
		if item.LogoURL != "" {
			element.LogoUrl = &item.LogoURL
		}
		if item.Tags != nil {
			element.Tags = item.Tags
		}
		if item.TechnologyLinks != nil {
			element.TechnologyLinks = item.TechnologyLinks
		}
		if item.Repo != "" {
			element.Repo = &item.Repo
		}
		if item.Branch != "" {
			element.Branch = &item.Branch
		}
		if item.FilePath != "" {
			element.FilePath = &item.FilePath
		}
		if item.Language != "" {
			element.Language = &item.Language
		}
		if item.BypassNoiseGate != nil {
			element.BypassNoiseGate = item.BypassNoiseGate
		}
		if item.HasView != nil {
			element.HasView = *item.HasView
		}
		if item.ViewLabel != "" {
			element.ViewLabel = &item.ViewLabel
		}
		if item.X != nil && item.Y != nil {
			placement := &diagv1.PlanViewPlacement{ParentRef: "root", PositionX: item.X, PositionY: item.Y}
			if len(element.GetPlacements()) > 0 {
				placement = element.GetPlacements()[0]
				if placement.ParentRef == "" {
					placement.ParentRef = "root"
				}
				placement.PositionX = item.X
				placement.PositionY = item.Y
			}
			element.Placements = []*diagv1.PlanViewPlacement{placement}
		}
	}

	if len(explicitRefs) > 0 {
		for _, connector := range parsed.Connectors {
			if next := explicitRefs[connector.GetSourceElementRef()]; next != "" {
				connector.SourceElementRef = next
			}
			if next := explicitRefs[connector.GetTargetElementRef()]; next != "" {
				connector.TargetElementRef = next
			}
			if next := explicitRefs[connector.GetViewRef()]; next != "" {
				connector.ViewRef = next
			}
		}
	}

	for index, connector := range parsed.Connectors {
		if index >= len(metadata.Connectors) {
			return
		}
		item := metadata.Connectors[index]
		if item.Ref != "" {
			connector.Ref = item.Ref
		}
		if item.SourceRef != "" {
			connector.SourceElementRef = item.SourceRef
		}
		if item.TargetRef != "" {
			connector.TargetElementRef = item.TargetRef
		}
		if item.Label != "" {
			connector.Label = &item.Label
		}
		if item.Description != "" {
			connector.Description = &item.Description
		}
		if item.Relationship != "" {
			connector.Relationship = &item.Relationship
		}
		if item.Direction != "" {
			connector.Direction = &item.Direction
		}
		if item.Style != "" {
			connector.Style = &item.Style
		}
		if item.URL != "" {
			connector.Url = &item.URL
		}
		if item.SourceHandle != "" {
			connector.SourceHandle = &item.SourceHandle
		}
		if item.TargetHandle != "" {
			connector.TargetHandle = &item.TargetHandle
		}
	}
}

func EscapeMetadataValue(value string) string {
	replacer := strings.NewReplacer(
		`\`, `\\`,
		"\n", `\n`,
		"\r", `\r`,
		"\t", `\t`,
		"=", `\=`,
		",", `\,`,
		"|", `\|`,
		":", `\:`,
	)
	return replacer.Replace(value)
}

func UnescapeMetadataValue(value string) (string, bool) {
	return unescapeMetadataValue(value)
}

func EncodeStringList(values []string) string {
	encoded := make([]string, 0, len(values))
	for _, value := range values {
		if strings.TrimSpace(value) == "" {
			continue
		}
		encoded = append(encoded, EscapeMetadataValue(value))
	}
	return strings.Join(encoded, ",")
}

func DecodeStringList(value string) ([]string, bool) {
	return decodeStringList(value)
}

func EncodeTechnologyLinks(values []*diagv1.TechnologyLink) string {
	encoded := make([]string, 0, len(values))
	for _, link := range values {
		if strings.TrimSpace(link.GetLabel()) == "" {
			continue
		}
		primary := "0"
		if link.GetIsPrimaryIcon() {
			primary = "1"
		}
		encoded = append(encoded, strings.Join([]string{
			EscapeMetadataValue(link.GetType()),
			EscapeMetadataValue(link.GetSlug()),
			EscapeMetadataValue(link.GetLabel()),
			primary,
		}, ":"))
	}
	return strings.Join(encoded, "|")
}

func parseMetadataPairs(text string) map[string]string {
	matches := metadataPairPattern.FindAllStringSubmatchIndex(text, -1)
	if len(matches) == 0 {
		return nil
	}
	pairs := map[string]string{}
	for index, match := range matches {
		keyStart := match[4]
		keyEnd := match[5]
		valueStart := keyEnd + 1
		valueEnd := len(text)
		if index+1 < len(matches) {
			valueEnd = matches[index+1][0]
		}
		pairs[text[keyStart:keyEnd]] = text[valueStart:valueEnd]
	}
	return pairs
}

func parseElementMetadata(pairs map[string]string) elementMetadata {
	item := elementMetadata{}
	stringKeys := map[string]*string{
		"ref":       &item.Ref,
		"kind":      &item.Kind,
		"desc":      &item.Description,
		"tech":      &item.Technology,
		"url":       &item.URL,
		"logo":      &item.LogoURL,
		"repo":      &item.Repo,
		"branch":    &item.Branch,
		"file":      &item.FilePath,
		"lang":      &item.Language,
		"viewLabel": &item.ViewLabel,
	}
	for key, target := range stringKeys {
		if raw, ok := pairs[key]; ok {
			if decoded, ok := unescapeMetadataValue(raw); ok && decoded != "" {
				*target = decoded
			}
		}
	}
	if raw, ok := pairs["x"]; ok {
		if decoded, ok := decodeNumber(raw); ok {
			item.X = &decoded
		}
	}
	if raw, ok := pairs["y"]; ok {
		if decoded, ok := decodeNumber(raw); ok {
			item.Y = &decoded
		}
	}
	if raw, ok := pairs["tags"]; ok {
		if decoded, ok := decodeStringList(raw); ok && len(decoded) > 0 {
			item.Tags = decoded
		}
	}
	if raw, ok := pairs["techLinks"]; ok {
		if decoded, ok := decodeTechnologyLinks(raw); ok && len(decoded) > 0 {
			item.TechnologyLinks = decoded
		}
	}
	if raw, ok := pairs["bypass"]; ok {
		if decoded, ok := decodeBool(raw); ok {
			item.BypassNoiseGate = &decoded
		}
	}
	if raw, ok := pairs["hasView"]; ok {
		if decoded, ok := decodeBool(raw); ok {
			item.HasView = &decoded
		}
	}
	return item
}

func parseConnectorMetadata(pairs map[string]string) connectorMetadata {
	item := connectorMetadata{}
	stringKeys := map[string]*string{
		"ref":          &item.Ref,
		"source":       &item.SourceRef,
		"target":       &item.TargetRef,
		"label":        &item.Label,
		"desc":         &item.Description,
		"rel":          &item.Relationship,
		"dir":          &item.Direction,
		"style":        &item.Style,
		"url":          &item.URL,
		"sourceHandle": &item.SourceHandle,
		"targetHandle": &item.TargetHandle,
	}
	for key, target := range stringKeys {
		if raw, ok := pairs[key]; ok {
			if decoded, ok := unescapeMetadataValue(raw); ok && decoded != "" {
				*target = decoded
			}
		}
	}
	return item
}

func mergeConnectorMetadata(base, override connectorMetadata) connectorMetadata {
	if override.Ref != "" {
		base.Ref = override.Ref
	}
	if override.SourceRef != "" {
		base.SourceRef = override.SourceRef
	}
	if override.TargetRef != "" {
		base.TargetRef = override.TargetRef
	}
	if override.Label != "" {
		base.Label = override.Label
	}
	if override.Description != "" {
		base.Description = override.Description
	}
	if override.Relationship != "" {
		base.Relationship = override.Relationship
	}
	if override.Direction != "" {
		base.Direction = override.Direction
	}
	if override.Style != "" {
		base.Style = override.Style
	}
	if override.URL != "" {
		base.URL = override.URL
	}
	if override.SourceHandle != "" {
		base.SourceHandle = override.SourceHandle
	}
	if override.TargetHandle != "" {
		base.TargetHandle = override.TargetHandle
	}
	return base
}

func unescapeMetadataValue(value string) (string, bool) {
	var out strings.Builder
	for index := 0; index < len(value); index++ {
		if value[index] != '\\' {
			out.WriteByte(value[index])
			continue
		}
		index++
		if index >= len(value) {
			return "", false
		}
		switch value[index] {
		case '\\', '=', ',', '|', ':':
			out.WriteByte(value[index])
		case 'n':
			out.WriteByte('\n')
		case 'r':
			out.WriteByte('\r')
		case 't':
			out.WriteByte('\t')
		default:
			return "", false
		}
	}
	return out.String(), true
}

func splitEscaped(value string, separator byte) []string {
	parts := []string{}
	var current strings.Builder
	for index := 0; index < len(value); index++ {
		char := value[index]
		if char == '\\' && index+1 < len(value) {
			current.WriteByte(char)
			current.WriteByte(value[index+1])
			index++
			continue
		}
		if char == separator {
			parts = append(parts, current.String())
			current.Reset()
			continue
		}
		current.WriteByte(char)
	}
	parts = append(parts, current.String())
	return parts
}

func decodeStringList(value string) ([]string, bool) {
	if value == "" {
		return []string{}, true
	}
	out := []string{}
	for _, raw := range splitEscaped(value, ',') {
		item, ok := unescapeMetadataValue(raw)
		if !ok {
			return nil, false
		}
		out = append(out, item)
	}
	return out, true
}

func decodeTechnologyLinks(value string) ([]*diagv1.TechnologyLink, bool) {
	if value == "" {
		return []*diagv1.TechnologyLink{}, true
	}
	out := []*diagv1.TechnologyLink{}
	for _, raw := range splitEscaped(value, '|') {
		fields := splitEscaped(raw, ':')
		if len(fields) < 3 || len(fields) > 4 {
			return nil, false
		}
		linkType, ok := unescapeMetadataValue(fields[0])
		if !ok {
			return nil, false
		}
		slug, ok := unescapeMetadataValue(fields[1])
		if !ok {
			return nil, false
		}
		label, ok := unescapeMetadataValue(fields[2])
		if !ok || strings.TrimSpace(label) == "" {
			return nil, false
		}
		primary := "0"
		if len(fields) == 4 {
			primary, ok = unescapeMetadataValue(fields[3])
			if !ok {
				return nil, false
			}
		}
		if linkType != "catalog" && linkType != "custom" {
			return nil, false
		}
		out = append(out, &diagv1.TechnologyLink{
			Type:          linkType,
			Slug:          ptrString(slug),
			Label:         label,
			IsPrimaryIcon: primary == "1" || primary == "true",
		})
	}
	return out, true
}

func decodeNumber(value string) (float64, bool) {
	raw, ok := unescapeMetadataValue(value)
	if !ok {
		return 0, false
	}
	number, err := strconv.ParseFloat(raw, 64)
	return number, err == nil
}

func decodeBool(value string) (bool, bool) {
	raw, ok := unescapeMetadataValue(value)
	if !ok {
		return false, false
	}
	switch raw {
	case "1", "true":
		return true, true
	case "0", "false":
		return false, true
	default:
		return false, false
	}
}

func exportedFlowchartElementRef(line string) string {
	match := exportedFlowchartElementRefLineRe.FindStringSubmatch(line)
	if match == nil {
		return ""
	}
	return match[1]
}
